src/pal/src/misc/miscpalapi.cpp

```diff
@@ -226,86 +234,162 @@ PAL_GetPALDirectoryA(
+    while (offset != dwLength);
+
+    _ASSERTE(offset == dwLength);
+
+    close(rand_des);
+
+    return TRUE;
+}
+
+#if HAVE_SYS_GETRANDOM
+static
+BOOL
+PAL_RandomFromGetRandom(
+        IN OUT LPVOID lpBuffer,
+        IN DWORD dwLength)
+{
+    ssize_t length = (ssize_t)dwLength;
+    static BOOL sMissingGetRandom;
+
+    if (sMissingGetRandom)
+    {
+        return FALSE;
+    }
+
+    do
+    {
+        ssize_t res = syscall(SYS_getrandom, lpBuffer, length,
+                              GRND_NONBLOCK);
+        if (res == -1)
         {
-            DWORD offset = 0;
-            do
+            if (errno == ENOSYS)
             {
-                ssize_t n = read(rand_des, (BYTE*)lpBuffer + offset , dwLength - offset);
-                if (n == -1)
-                {
-                    if (errno == EINTR)
-                    {
-                        continue;
-                    }
-                    ASSERT("read() failed, errno:%d (%s)\n", errno, strerror(errno));
-
-                    break;
-                }
-
-                offset += n;
+                sMissingGetRandom = TRUE;
+                return FALSE;
+            }
+            else if (errno == EINTR)
+            {
+                continue;
+            }
+            else
+            {
+                return FALSE;
             }
-            while (offset != dwLength);
-
-            _ASSERTE(offset == dwLength);
-
-            close(rand_des);
         }
+
+        length -= res;
+        lpBuffer = (char *)lpBuffer + res;
     }
+    while (length);
 
-    if (!sInitializedMRand)
+    return TRUE;
+}
+#endif // HAVE_SYS_GETRANDOM
+
+VOID
+PALAPI
+PAL_Random(
+        IN OUT LPVOID lpBuffer,
+        IN DWORD dwLength)
+{
+    static BOOL sInitializedMRand;
+    BOOL needSrand = TRUE;
+
+    PERF_ENTRY(PAL_Random);
+    ENTRY("PAL_Random(lpBuffer=%p, dwLength=%d)\n", lpBuffer, dwLength);
+
+#if HAVE_SYS_GETRANDOM
+    if (PAL_RandomFromGetRandom(lpBuffer, dwLength))
     {
-        srand48(time(NULL));
-        sInitializedMRand = TRUE;
+        needSrand = FALSE;
+    }
+    else
+#endif // HAVE_SYS_GETRANDOM
+    if (PAL_RandomFromUrandom(lpBuffer, dwLength))
+    {
+        needSrand = FALSE;
     }
 
-    // always xor srand48 over the whole buffer to get some randomness
-    // in case /dev/urandom is not really random
-
-    for (i = 0; i < dwLength; i++)
+    if (needSrand)
     {
-        if (i % sizeof(long) == 0) {
-            num = mrand48();
+        if (!sInitializedMRand)
+        {
+            // FIXME: There's not enough entropy in time(NULL) to
+            // initialize the PRNG.  Might be a good idea to leverage
+            // other sources of entropy at this point.
+            srand48(time(NULL));
+            sInitializedMRand = TRUE;
         }
 
-        *(((BYTE*)lpBuffer) + i) ^= num;
-        num >>= 8;
+        while (dwLength)
+        {
+            const auto num = mrand48();
+
+            if (dwLength >= sizeof(num))
```

Comment:
May I suggest a little simplification? Instead of the if / else here:
```c++
DWORD copySize = std::min(dwLength, sizeof(num));
memcpy(lpBuffer, &num, copySize);
lpBuffer = (char *)lpBuffer + copySize;
dwLength -= copySize;
```

Reply:
I can do this, but memcpy() with a known size at compile time generates slightly better code (if size fits in a register, it's often just a MOV when unaligned loads are OK in that platform).

Reply:
Right, that's true. Forget my suggestion then.

Reply:
It would be good to look at the generated code in release build to verify that the memcpy really got replaced by assignment.

Reply:
Here's a [godbolt link](https://gcc.godbolt.org/z/iMWcaq) with both versions.